### PR TITLE
Restore checkpoint in the background when current checkpoint is not old enough

### PR DESCRIPTION
**Pull request by JuanCarlosGI**
This should reduce startup time when a machine is being restarted.

### CODE REVIEW

File: Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs [hunk 1]
```diff
@@ -344,7 +347,14 @@ public Task<BoolResult> RestoreCheckpointAsync(OperationContext context, string
                             }
 
                             // Restoring the checkpoint
-                            return _database.RestoreCheckpoint(context, extractedCheckpointDirectory);
+                            var result = _database.RestoreCheckpoint(context, extractedCheckpointDirectory);
+
+                            if (result)
+                            {
```
Comment (dcwuser) [inline Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs]: Comments please on what this is for.

**Comment by SergeyTeplyakov** [comment]
>                         result = await RestoreCheckpointStateAsync(context, checkpointState);

Lets be explicit here and pass 'force: false'.

---
Refers to: Public/Src/Cache/ContentStore/Distributed/NuCache/LocalLocationStore.cs:429 in 373c47e. [](commit_id = 373c47ece75d8db3d5a27879d8ace0621d449095, deletion_comment = False)

File: Public/Src/Cache/ContentStore/Distributed/NuCache/LocalLocationStore.cs [hunk 2]
```diff
@@ -598,15 +598,34 @@ internal async Task<BoolResult> CreateCheckpointAsync(OperationContext context)
...
         {
             var token = context.Token;
+
+            if (!force)
+            {
+                var latestCheckpoint = _checkpointManager.GetLatestCheckpointInfo(context);
+                var latestCheckpointAge = _clock.UtcNow - latestCheckpoint?.checkpointTime;
+                var shouldRestoreInBackground = latestCheckpointAge < _configuration.Checkpoint.RestoreCheckpointAgeThreshold;
+
+                if (shouldRestoreInBackground)
```
Comment (SergeyTeplyakov) [inline Public/Src/Cache/ContentStore/Distributed/NuCache/LocalLocationStore.cs]: >shouldRestoreInBackground [](start = 20, length = 25)

Is it ok, that we trace here but still do the same thing regardless of the check?

Comment (SergeyTeplyakov) [inline Public/Src/Cache/ContentStore/Distributed/NuCache/LocalLocationStore.cs]: I think we need to explicitely trace that we skipped the eager checkpoint restore with the data about the checkpoint.

File: Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs [hunk 3]
```diff
@@ -356,6 +367,36 @@ public Task<BoolResult> RestoreCheckpointAsync(OperationContext context, string
...
 
+        private void WriteLatestCheckpointToFile(OperationContext context, CheckpointState checkpointState)
+        {
+            try
+            {
+                _fileSystem.WriteAllText(_lastCheckpointFile, $"{checkpointState.CheckpointId},{checkpointState.CheckpointTime}");
+            }
+            catch (Exception e)
+            {
+                _tracer.Debug(context, $"Failed to write latest checkpoint state to disk: {e}");
```
Comment (SergeyTeplyakov) [inline Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs]: nit: I think it should be a warning. Technically, this should never happen...

File: Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs [hunk 4]
```diff
@@ -356,6 +367,36 @@ public Task<BoolResult> RestoreCheckpointAsync(OperationContext context, string
...
+            }
+        }
+
+        public (string id, DateTime checkpointTime)? GetLatestCheckpointInfo(OperationContext context)
```
Comment (SergeyTeplyakov) [inline Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs]: >GetLatestCheckpointInfo [](start = 53, length = 23)

nit: checkpointId?

The naming looks a bit inconsistent. It should be either 'checkpointId and checkpointTime' or 'id' and 'time', imo.

**Pull request merged**